File: java/23_file_handling.java
// ============================================================
// Topic:  File Handling  (FOCUS: java.nio.file.Files, Paths, Path)
// Summary: write, read, append, exists, line count, delete.
// Run:     java java/23_file_handling.java   (Java 11+ single-file launch)
//
// KEY DIFFERENCES:
//   Python:  open()/read()/write(); pathlib.Path
//   TS/JS:   fs.readFileSync / fs.promises; path.join
//   Rust:    std::fs::File; std::path::Path
//   Java:    java.nio.file.Files + Path (modern NIO.2 API)
//   C++:     std::fstream; std::filesystem::path
// ============================================================

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.List;

public class FileHandling {

    public static void main(String[] args) throws IOException {
        // Demo file in the working directory (deleted at the end).
        Path path = Path.of("_alllangs_demo.txt");

        // ----- 1. Write text to a file -----
        Files.writeString(path, "hello file", StandardCharsets.UTF_8);
        System.out.println("write: ok");                                    // Output: write: ok

        // ----- 2. Read entire file as a String -----
        String content = Files.readString(path, StandardCharsets.UTF_8);
        System.out.println("read: " + content);                             // Output: read: hello file

        // ----- 3. Append to the file -----
        Files.writeString(
                path,
                "!",
                StandardCharsets.UTF_8,
                StandardOpenOption.APPEND
        );
        System.out.println("append: ok");                                 // Output: append: ok

        // ----- 4. Read again after append -----
        content = Files.readString(path, StandardCharsets.UTF_8);
        System.out.println("read: " + content);                             // Output: read: hello file!

        // ----- 5. Check whether the path exists -----
        boolean exists = Files.exists(path);
        System.out.println("exists: " + exists);                            // Output: exists: true

        // ----- 6. Count lines -----
        List<String> lines = Files.readAllLines(path, StandardCharsets.UTF_8);
        System.out.println("lines: " + lines.size());                       // Output: lines: 1

        // ----- 7. Delete the file -----
        Files.delete(path);
        System.out.println("deleted: ok");                                  // Output: deleted: ok

        // Cross-language note:
        //   Path.of("file.txt") == Python pathlib.Path("file.txt").
        //   Files.writeString / readString replace older FileWriter/BufferedReader.
        //   Paths.get(...) still works; Path.of is the Java 11+ shorthand.
        //   Always delete demo files so reruns start clean.
    }
}

File: csharp/23_file_handling.cs
// ============================================================
// Topic:  File Handling
// Summary: Write, read, append, exists check, line count, delete via System.IO.File.
// Run:     dotnet script csharp/23_file_handling.cs
//
// KEY DIFFERENCES:
//   C#:     System.IO.File / FileStream; Path.Combine for cross-platform paths
//   Python: open(), pathlib.Path; with-statement for auto-close
//   TS/JS:  fs.readFileSync / fs.promises; path.join
//   Rust:   std::fs; std::path::PathBuf
//   Java:   java.nio.file.Files, Paths.get
//   C++:    std::fstream; std::filesystem::path
// ============================================================

using System;
using System.IO;

// ----- 1. Temp path — cross-platform via Path.GetTempPath() -----
string path = Path.Combine(Path.GetTempPath(), "_alllangs_demo.txt");

// ----- 2. Write text to file -----
File.WriteAllText(path, "hello file");
Console.WriteLine("write: ok");                        // write: ok

// ----- 3. Read entire file -----
string content = File.ReadAllText(path);
Console.WriteLine($"read: {content}");                 // read: hello file

// ----- 4. Append to file -----
File.AppendAllText(path, "!");
Console.WriteLine("append: ok");                       // append: ok

// ----- 5. Read after append -----
content = File.ReadAllText(path);
Console.WriteLine($"read: {content}");                 // read: hello file!

// ----- 6. Check existence -----
Console.WriteLine($"exists: {File.Exists(path).ToString().ToLowerInvariant()}"); // exists: true

// ----- 7. Count lines -----
int lines = File.ReadAllLines(path).Length;
Console.WriteLine($"lines: {lines}");                    // lines: 1

// ----- 8. Delete file -----
File.Delete(path);
Console.WriteLine("deleted: ok");                      // deleted: ok